et_replay/comm/profiler_trace_analysis.py

```diff
@@ -0,0 +1,252 @@
+                   and i['name'].startswith('ncclDevKernel_') \
+                   and 'algbw (GB/sec)' in i['args']]
+    for evt in nccl_events:
+        knl_name = evt['name'][:evt['name'].index('(')]
+        data_size = _calculate_event_data_size(evt)
+        ranks_count = evt['args']['Group size']
+
+        ranks = ast.literal_eval(evt['args']['Process Group Ranks'])
+        pg_id = int(evt['args']['Process Group Name'])
+        pg = tuple([*ranks, pg_id]) if rank == min(ranks) else None
+
+        comm_bw_data[(knl_name, data_size, ranks_count)].append(
+            [evt['dur'], evt['args']['algbw (GB/sec)'], evt['args']['busbw (GB/sec)'], pg]
+        )
+
+def analyze_profiler_trace(trace_dir: str, report_dir: str):
+    '''
+    Analyse input PyTorch profiler trace (i.e. Kineto trace) and generate report.
+
+    Args:
+        trace_dir (str): dir path of input traces, where trace name should be in "rank-n.json" format.
+        report_dir (str): dir path for generated reports 
+    '''
+    logger.info(f'Parse profiler trace from "{trace_dir}" and generate reports to "{report_dir}"')
+
+    processed_trace_dir = os.path.join(report_dir, 'profiler_trace_processed')
+    pathlib.Path(processed_trace_dir).mkdir(parents=True, exist_ok=True)
+
+    # list of iteration time in all ranks
+    iter_e2e_time = []
+
+    # list of shared bw
+    sbw_lst = []
+
+    # key is (kernel_name, data size, ranks number)
+    # value is list of [dur, algbw, busbw, pg]
+    comm_bw_data = defaultdict(list)
+
+    for fpath in os.scandir(trace_dir):
+        if not fpath.is_file():
+            continue
+
+        with open(fpath.path, 'r', encoding='utf-8') as f:
+            trace = json.load(f)
+
+        calculate_bw_(trace)
+        with open(os.path.join(processed_trace_dir, fpath.name), 'w', encoding='utf-8') as f:
+            json.dump(trace, f)
+        
+        sbw_lst.append(calculate_sbw(trace))
+
+        pick_iter_e2e_time_(trace, iter_e2e_time)
+        pick_comm_bw_(trace, comm_bw_data)
+    
+    comm_bw_summary = {}
+    for k,v in comm_bw_data.items():
+        t_lst     = [i[0] for i in v]
+        busbw_lst = [i[2] for i in v]
+        pg_set    = set([i[3] for i in v if i[3]])
+        comm_bw_summary[k] = [len(pg_set),
+                              np.average(t_lst), 
+                              np.average(busbw_lst),
+                              np.percentile(busbw_lst,  1),
+                              np.percentile(busbw_lst, 50),
+                              np.percentile(busbw_lst, 90),
+                              np.percentile(busbw_lst, 99)]
+    comm_bw_summary = dict(sorted(comm_bw_summary.items()))
+
+    # dump summary report
+    with open(os.path.join(report_dir, 'profiler_trace_summary_report.txt'), 'w', encoding='utf-8') as f:
+        f.write(f'avg. E2ETime of iters among all ranks: {sum(iter_e2e_time) / len(iter_e2e_time) / 1e3 :.3f} ms\n')
```

Comment:
You can consider using tabulate library since you are considering adding dependency.